src/Dungeon/DungeonGenerator.elm

```diff
@@ -1,358 +1,573 @@
+         2.  A bend with the bend either being 45deg or 90deg.
+             eg from (0,0) to (2, 3) there are 4 possible outcomes with one bend ending at (2, 3)
+             a. along the y (0, 1) and (0,3)
+             b. along the x (2, 0) and (2, 2)
+
+-}
+possiblePaths : List Direction -> Vector -> Vector -> List Vector
+possiblePaths validDirections (( x_a, y_a ) as a) (( x_b, y_b ) as b) =
     let
-        corridorBoundary =
-            Corridor.boundary corridor
+        (( d_x, d_y ) as d_vector) =
+            ( abs (x_b - x_a), abs (y_b - y_a) )
+
+        pathFromDirection direction =
+            let
+                alongAxis =
+                    Vector.mul (Vector.fromDirection direction) d_vector
+                        |> Vector.add a
 
-        withinBounds =
-            List.all (\x -> Config.withinDungeonBounds x model.config) corridorBoundary
+                diagonal =
+                    Vector.scaleInt (min d_x d_y) (Vector.fromDirection direction)
+                        |> Vector.add a
 
-        overlapping =
-            List.any (\x -> Set.member x model.occupied) corridorBoundary
+                diagonalToB =
+                    Vector.sub b diagonal
+                        |> Vector.add a
+            in
+            if Direction.isCardinal direction then
+                [ alongAxis ]
+            else
+                [ diagonal, diagonalToB ]
     in
-    not overlapping && withinBounds
+    List.concatMap pathFromDirection validDirections
+
+
+
+--------------
+-- Cleaning --
+--------------
 
 
-canFitRoom : Model -> Room -> Bool
-canFitRoom model room =
+clean : Dungeon -> Dungeon
+clean dungeon =
     let
-        roomPositions =
-            Room.floorsAndBoundaries room
+        replaceTile tile dict =
+            calculateTypeOfWall dungeon.map tile.position
+                |> (\newTile -> Dict.insert newTile.position newTile dict)
+    in
+    Dict.values dungeon.map
+        |> List.foldl replaceTile dungeon.map
+        |> (\newMap -> { dungeon | map = newMap })
 
-        withinBounds =
-            List.all (\x -> Config.withinDungeonBounds x model.config) roomPositions
 
-        overlapping =
-            List.any (\x -> Set.member x model.occupied) roomPositions
+calculateTypeOfWall : Dict Vector Tile -> Vector -> Tile
+calculateTypeOfWall map position =
+    case ( hasAdjacentFloors position map, hasThreeOrMoreNeighbourFloors position map ) of
+        ( True, True ) ->
+            Tile.toTile position Tile.Types.DarkDgn
+
+        ( True, False ) ->
+            Tile.toTile position Tile.Types.WallDarkDgn
+
+        _ ->
+            Tile.toTile position Tile.Types.Rock
+
+
+adjacentNeighbourPairs : List (List Direction)
+adjacentNeighbourPairs =
+    [ [ N, E ]
+    , [ E, S ]
+    , [ S, W ]
+    , [ W, N ]
+    ]
+
+
+adjacentNeighbourTriplets : List (List Direction)
+adjacentNeighbourTriplets =
+    [ [ N, E, S ]
+    , [ E, S, W ]
+    , [ S, W, N ]
+    , [ W, N, E ]
+    ]
```

Comment:
I wonder if this would show it better:




```
let
  directions = 
    [ N, E, S, W ]
  directionsWithout directionToFilter =
    List.filter (\direction -> direction \= directionToFilter)
in
  List.map directionsWithout directions
```

This is a Haskell version:

```
map directionsWithout directions
  where
    directions = [ N, E, S, W ]
    directionsWithout direction = filter (\= direction) directions
```

Reply:
Often I go to refactor things in Elm, and then I realise half way through that they're going to be less terse than i'd hoped.